pricegraph/src/orderbook/scalar.rs

```diff
@@ -169,3 +170,52 @@ fn assert_strictly_positive_and_finite(value: f64) -> f64 {
     debug_assert!(num::is_strictly_positive_and_finite(value));
     value
 }
+
+/// The exchange rate weight used by the pathfinding algorithm.
+#[derive(Clone, Copy, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Weight(i128);
+
+impl Weight {
+    /// Creates a new graph weight from a floating point number.
+    pub fn new(value: f64) -> Self {
+        // NOTE: We need to reserve some bits in order for the addition not to
+        // overflow, specifically we need to reserve
+        // - 16 bits for the maximum number of nodes (maximum number of tokens)
+        // - 8 bits for the [-128, 128) range
+        // - 1 bit for infinity
+        // Thats a total of 25 bits, which leaves us with 103 bits of precision
+        // for the weight.
+        const FACTOR: f64 = (1u128 << 103) as _;
+
+        let weight = value.log2() * FACTOR;
+        debug_assert!(-128.0 * FACTOR <= weight && weight < 128.0 * FACTOR);
+
+        Weight(weight as _)
```

Comment:
Thinking more about precision, it might be possible to have less error by creating our log2 fixed point value directly from the u128 numerator and denominator instead of calculating the ratio as f64 (in LimitPrice and ExchangeRate) first and then moving into fixed point.

Reply:
This is a really good idea, but I'd have to research how. Maybe this is provided by the fixed point crate?

Reply:
Agree, can be done later.